Quarkus REST client is unable to determine base URI with MP-style properties after MP bump and I can't see anything documented

### Describe the bug

My app started failing after https://github.com/quarkusio/quarkus/pull/43959. There is nothing in migration guide and I don't know if this is expected. Please provide context. Thanks

### Expected behavior

Documentation or app works. Worked with 3.15.x.

### Actual behavior

REST client is missing base URI after MP bump.

```
2024-10-24 10:16:37,494 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /client-hello failed, error id: 110a707e-daa0-420f-9843-50bfc2bae118-1: java.lang.RuntimeException: Error injecting org.acme.GreetingsService org.acme.GreetingClientResource.greetingsService
	at org.acme.GreetingClientResource_Bean.doCreate(Unknown Source)
	at org.acme.GreetingClientResource_Bean.create(Unknown Source)
	at org.acme.GreetingClientResource_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:32)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.ComputingCacheContextInstances.computeIfAbsent(ComputingCacheContextInstances.java:19)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
	at org.acme.GreetingClientResource_Bean.get(Unknown Source)
	at org.acme.GreetingClientResource_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:559)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:539)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:572)
	at io.quarkus.arc.impl.ArcContainerImpl$3.get(ArcContainerImpl.java:331)
	at io.quarkus.arc.impl.ArcContainerImpl$3.get(ArcContainerImpl.java:328)
	at io.quarkus.arc.runtime.BeanContainerImpl$1.create(BeanContainerImpl.java:58)
	at io.quarkus.resteasy.reactive.common.runtime.ArcBeanFactory.createInstance(ArcBeanFactory.java:27)
	at org.jboss.resteasy.reactive.server.handlers.InstanceHandler.handle(InstanceHandler.java:26)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:139)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:627)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalArgumentException: Unable to determine the proper baseUrl/baseUri. Consider registering using @RegisterRestClient(baseUri="someuri"), @RegisterRestClient(configKey="orkey"), or by adding 'quarkus.rest-client."org.acme.GreetingsService".url' or 'quarkus.rest-client."org.acme.GreetingsService".uri' to your Quarkus configuration
	at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.configureBaseUrl(RestClientCDIDelegateBuilder.java:379)
	at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.configureBuilder(RestClientCDIDelegateBuilder.java:76)
	at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.build(RestClientCDIDelegateBuilder.java:71)
	at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.createDelegate(RestClientCDIDelegateBuilder.java:52)
	at io.quarkus.rest.client.reactive.runtime.RestClientReactiveCDIWrapperBase.delegate(RestClientReactiveCDIWrapperBase.java:76)
	at io.quarkus.rest.client.reactive.runtime.RestClientReactiveCDIWrapperBase.<init>(RestClientReactiveCDIWrapperBase.java:30)
	at org.acme.GreetingsService$$CDIWrapper.<init>(Unknown Source)
	at org.acme.GreetingsService$$CDIWrapper_ClientProxy.<init>(Unknown Source)
	at org.acme.GreetingsService$$CDIWrapper_Bean.proxy(Unknown Source)
	at org.acme.GreetingsService$$CDIWrapper_Bean.get(Unknown Source)
	at org.acme.GreetingsService$$CDIWrapper_Bean.get(Unknown Source)
	... 31 more
```

### How to Reproduce?

Steps to reproduce the behavior: 

```
quarkus create app --extensions=rest-client,quarkus-rest
cd code-with-quarkus/

cat <<EOF > src/main/java/org/acme/GreetingsService.java
package org.acme;

import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

@RegisterRestClient
public interface GreetingsService {

    @Path("/hello")
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    String hello();
    
}
EOF

cat <<EOF > src/main/java/org/acme/GreetingClientResource.java
package org.acme;

import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import org.eclipse.microprofile.rest.client.inject.RestClient;

@Path("/client-hello")
public class GreetingClientResource {
    
    @RestClient
    @Inject
    GreetingsService greetingsService;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return greetingsService.hello();
    }
}
EOF

cat <<EOF > src/test/java/org/acme/GreetingClientResourceTest.java
package org.acme;

import io.quarkus.test.junit.QuarkusTest;
import org.junit.jupiter.api.Test;

import static io.restassured.RestAssured.given;
import static org.hamcrest.CoreMatchers.is;

@QuarkusTest
class GreetingClientResourceTest {
    @Test
    void testHelloEndpoint() {
        given()
          .when().get("/client-hello")
          .then()
             .statusCode(200)
             .body(is("Hello from Quarkus REST"));
    }

}
EOF

cat <<EOF > src/main/resources/application.properties
GreetingsService/mp-rest/url=http://localhost:8081
GreetingsService/mp-rest/scope=jakarta.inject.Singleton
EOF


PASS: mvn clean test
FAIL: mvn clean test -Dquarkus.platform.version=999-SNAPSHOT -Dquarkus.platform.group-id=io.quarkus
```

### Output of `uname -a` or `ver`

Fedora

### Output of `java -version`

Temurin 21

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.9.4

### Additional information

_No response_

Fun fact @radcortez , setting `@RegisterRestClient(configKey = "GreetingsService")` leads to:

```
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.rest.client.reactive.deployment.devservices.DevServicesRestClientHttpProxyProcessor#determineRequiredProxies threw an exception: io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.lang.IllegalStateException: Too many recursive interceptor actions
	java.lang.IllegalStateException: Too many recursive interceptor actions
	java.lang.IllegalStateException: Too many recursive interceptor actions
	at io.smallrye.config.SmallRyeConfig.buildMappings(SmallRyeConfig.java:139)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:93)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:771)
	at io.quarkus.restclient.config.RestClientsBuildTimeConfig.get(RestClientsBuildTimeConfig.java:121)
	at io.quarkus.rest.client.reactive.deployment.devservices.DevServicesRestClientHttpProxyProcessor.determineRequiredProxies(DevServicesRestClientHttpProxyProcessor.java:63)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)

	at io.quarkus.builder.Execution.run(Execution.java:124)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:161)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:350)
	... 6 more
Caused by: io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.lang.IllegalStateException: Too many recursive interceptor actions
	java.lang.IllegalStateException: Too many recursive interceptor actions
	java.lang.IllegalStateException: Too many recursive interceptor actions
	at io.smallrye.config.SmallRyeConfig.buildMappings(SmallRyeConfig.java:139)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:93)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:771)
	at io.quarkus.restclient.config.RestClientsBuildTimeConfig.get(RestClientsBuildTimeConfig.java:121)
	at io.quarkus.rest.client.reactive.deployment.devservices.DevServicesRestClientHttpProxyProcessor.determineRequiredProxies(DevServicesRestClientHttpProxyProcessor.java:63)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)

```

But linked Eclipse docs says I can use it. Did I misread it?

--- a/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/AbstractRestClientConfigBuilder.java
+++ b/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/AbstractRestClientConfigBuilder.java
@@ -74,19 +74,22 @@ public abstract class AbstractRestClientConfigBuilder implements ConfigBuilder {
             quarkusFallbacks.put(restClient.getSimpleName(), quotedSimpleName);
             relocates.put(quotedSimpleName, quotedFullName);
 
-            if (restClient.getConfigKey() != null) {
-                String quotedConfigKey = "\"" + restClient.getConfigKey() + "\"";
-                if (restClient.isConfigKeyComposed()) {
-                    // Quoted Simple Name -> Quoted Config Key
-                    quarkusFallbacks.put(quotedSimpleName, quotedConfigKey);
-                    relocates.put(quotedConfigKey, quotedFullName);
-                } else {
-                    // Quoted Simple Name -> Config Key
-                    quarkusFallbacks.put(quotedSimpleName, restClient.getConfigKey());
-                    relocates.put(restClient.getConfigKey(), quotedFullName);
-                    // Config Key -> Quoted Config Key
-                    quarkusFallbacks.put(restClient.getConfigKey(), quotedConfigKey);
-                    relocates.put(quotedConfigKey, quotedFullName);
+            String configKey = restClient.getConfigKey();
+            if (configKey != null && !restClient.isConfigKeyEqualsNames()) {
+                String quotedConfigKey = "\"" + configKey + "\"";
+                if (!quotedConfigKey.equals(quotedFullName) && !quotedConfigKey.equals(quotedSimpleName)) {
+                    if (restClient.isConfigKeyComposed()) {
+                        // Quoted Simple Name -> Quoted Config Key
+                        quarkusFallbacks.put(quotedSimpleName, quotedConfigKey);
+                        relocates.put(quotedConfigKey, quotedFullName);
+                    } else {
+                        // Quoted Simple Name -> Config Key
+                        quarkusFallbacks.put(quotedSimpleName, configKey);
+                        relocates.put(configKey, quotedFullName);
+                        // Config Key -> Quoted Config Key
+                        quarkusFallbacks.put(configKey, quotedConfigKey);
+                        relocates.put(quotedConfigKey, quotedFullName);
+                    }
                 }
             }
 
@@ -94,8 +97,8 @@ public abstract class AbstractRestClientConfigBuilder implements ConfigBuilder {
             String mpRestFullName = restClient.getFullName() + "/mp-rest/";
             microProfileFallbacks.put(quotedFullName, mpRestFullName);
             relocates.put(mpRestFullName, quotedFullName);
-            if (restClient.getConfigKey() != null) {
-                String mpConfigKey = restClient.getConfigKey() + "/mp-rest/";
+            if (configKey != null && !restClient.isConfigKeyEqualsNames()) {
+                String mpConfigKey = configKey + "/mp-rest/";
                 microProfileFallbacks.put(mpRestFullName, mpConfigKey);
                 relocates.put(mpConfigKey, quotedFullName);
             }

--- a/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RegisteredRestClient.java
+++ b/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RegisteredRestClient.java
@@ -35,6 +35,14 @@ public class RegisteredRestClient {
         return configKey;
     }
 
+    public boolean isConfigKeyEqualsNames() {
+        if (configKey == null) {
+            return false;
+        }
+
+        return configKey.equals(fullName) || configKey.equals(simpleName);
+    }
+
     public boolean isConfigKeyComposed() {
         if (configKey == null) {
             throw new IllegalStateException("configKey is null");

--- a/extensions/resteasy-classic/rest-client-config/runtime/src/test/java/io/quarkus/restclient/config/RestClientConfigTest.java
+++ b/extensions/resteasy-classic/rest-client-config/runtime/src/test/java/io/quarkus/restclient/config/RestClientConfigTest.java
@@ -2,6 +2,7 @@ package io.quarkus.restclient.config;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.math.BigInteger;
@@ -65,6 +66,43 @@ class RestClientConfigTest {
         assertEquals(1, restClientsConfig.clients().size());
         assertTrue(restClientsConfig.clients().containsKey(ConfigKeyRestClient.class.getName()));
         verifyConfig(restClientsConfig.getClient(ConfigKeyRestClient.class));
+    }
+
+    @Test
+    void restClientConfigKeyMatchName() {
+        SmallRyeConfig config = ConfigUtils.emptyConfigBuilder()
+                .withMapping(RestClientsConfig.class)
+                .withCustomizers(new SmallRyeConfigBuilderCustomizer() {
+                    @Override
+                    public void configBuilder(final SmallRyeConfigBuilder builder) {
+                        new AbstractRestClientConfigBuilder() {
+                            @Override
+                            public List<RegisteredRestClient> getRestClients() {
+                                return List.of(new RegisteredRestClient(ConfigKeyRestClient.class,
+                                        ConfigKeyRestClient.class.getName()));
+                            }
+                        }.configBuilder(builder);
+                    }
+                })
+                .build();
+        assertNotNull(config);
+
+        config = ConfigUtils.emptyConfigBuilder()
+                .withMapping(RestClientsConfig.class)
+                .withCustomizers(new SmallRyeConfigBuilderCustomizer() {
+                    @Override
+                    public void configBuilder(final SmallRyeConfigBuilder builder) {
+                        new AbstractRestClientConfigBuilder() {
+                            @Override
+                            public List<RegisteredRestClient> getRestClients() {
+                                return List.of(new RegisteredRestClient(ConfigKeyRestClient.class,
+                                        ConfigKeyRestClient.class.getSimpleName()));
+                            }
+                        }.configBuilder(builder);
+                    }
+                })
+                .build();
+        assertNotNull(config);
     }
 
     @Test
@@ -147,6 +185,43 @@ class RestClientConfigTest {
         assertTrue(clientConfig.proxyAddress().isPresent());
         assertTrue(clientConfig.queryParamStyle().isPresent());
         assertThat(clientConfig.queryParamStyle().get()).isEqualTo(QueryParamStyle.COMMA_SEPARATED);
+    }
+
+    @Test
+    void restClientMicroProfileConfigKeyMatchName() {
+        SmallRyeConfig config = ConfigUtils.emptyConfigBuilder()
+                .withMapping(RestClientsConfig.class)
+                .withCustomizers(new SmallRyeConfigBuilderCustomizer() {
+                    @Override
+                    public void configBuilder(final SmallRyeConfigBuilder builder) {
+                        new AbstractRestClientConfigBuilder() {
+                            @Override
+                            public List<RegisteredRestClient> getRestClients() {
+                                return List.of(new RegisteredRestClient(MPConfigKeyRestClient.class,
+                                        MPConfigKeyRestClient.class.getName()));
+                            }
+                        }.configBuilder(builder);
+                    }
+                })
+                .build();
+        assertNotNull(config);
+
+        config = ConfigUtils.emptyConfigBuilder()
+                .withMapping(RestClientsConfig.class)
+                .withCustomizers(new SmallRyeConfigBuilderCustomizer() {
+                    @Override
+                    public void configBuilder(final SmallRyeConfigBuilder builder) {
+                        new AbstractRestClientConfigBuilder() {
+                            @Override
+                            public List<RegisteredRestClient> getRestClients() {
+                                return List.of(new RegisteredRestClient(MPConfigKeyRestClient.class,
+                                        MPConfigKeyRestClient.class.getSimpleName()));
+                            }
+                        }.configBuilder(builder);
+                    }
+                })
+                .build();
+        assertNotNull(config);
     }
 
     private void verifyConfig(RestClientConfig config) {

